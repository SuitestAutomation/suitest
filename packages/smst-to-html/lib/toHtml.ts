/// <reference path="../../smst/types/unistTestLine.d.ts" />
import {assertUnknownSectionNode} from '@suitest/smst/commonjs/jsxFactory';
import {escapeHtml} from './utils';

/**
 * Helper function to build HTML figure element
 * @param className - class name to be added to figure
 * @param content - body content of the figure, already renderer to string
 * @param titleNodes - optional figcaption
 * @param prefix - text to insert before figcaption
 */
const renderFigure = (className: string, content: string, titleNodes?: InlineTextNode[], prefix = ''): string => {
	const output = [`<figure class="${className}">`];

	if (titleNodes?.length) {
		output.push(`<figcaption>${prefix + titleNodes.map(renderNode).join('')}</figcaption>`);
	}

	output.push(content, '</figure>');

	return output.join('');
};

const renderHtmlCodeBlockNode = (node: CodeBlockNode): string => renderFigure(
	'suitest-test-line__code-block',
	`<pre><code>${node.value}</code></pre>`
);

const renderHtmlPropsNode = (node: PropertiesNode): string => renderFigure(
	'suitest-test-line__table',
	[
		'<table>',
		...node.children.map(prop => ([
			'<tr>',
				'<td>' + prop.name.map(renderNode).join('') + '</td>',
				'<td>' + prop.comparator + '</td>',
				'<td>' + toHtml(prop.expectedValue) + '</td>',
			'</tr>',
		].join(''))),
		'</table>',
	].join('')
);

const renderHtmlConditionNode = (node: ConditionNode): string => renderFigure(
	'suitest-test-line__condition',
	node.children.map(renderNode).join(''),
	node.title,
	'condition: '
);

const renderHtmlTestLineNode = (node: TestLineNode): string => renderFigure(
	'suitest-test-line',
	node.children.map(renderNode).join(''),
	node.title
);

const renderNode = (node: SingleNode): string => {
	switch (node.type) {
		case 'text':
			return escapeHtml(node.value ?? '');
		case 'code':
			return `<code class="suitest-test-line__text--code">${node.value ?? ''}</code>`;
		case 'subject':
		case 'input':
			return `<span class="suitest-test-line__text--${node.type}">${escapeHtml(node.value ?? '')}</span>`;
		case 'code-block':
			return renderHtmlCodeBlockNode(node);
		case 'props':
			return renderHtmlPropsNode(node);
		case 'prop':
			throw new Error('Property node should not be rendered outside of Properties collection');
		case 'test-line':
			return renderHtmlTestLineNode(node);
		case 'condition':
			return renderHtmlConditionNode(node);
		case 'test-line-result':
			return '';
		default:
			/* istanbul ignore next */
			return assertUnknownSectionNode(node);
	}
};

export const toHtml = (node: Node): string => {
	if (!Array.isArray(node)) {
		node = [node];
	}

	return node.map(renderNode).join('');
};
