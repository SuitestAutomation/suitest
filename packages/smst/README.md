# SMST

**S**uitest **m**essage **s**yntax **t**ree.

Suitest flavour of the [unist][ext-unist] domain-specific syntax tree, designed
to express test line definitions, results and other user-facing messages. SMST can
be used to render messages in plain text, text with ANSI styling and HTML.

## Nodes

### `Text`

Inline plain text.

```idl
interface Text {
    type: "text"
    value: string
}
```

### `Subject`

Inline text, that represents test line subject name

```idl
interface Subject <: Text {
    type: "subject"
}
```

See [Text][dfn-text].

### `Input`

Inline text, that represents user-defined values in test lines, e.g. expectations.

```idl
interface Input <: Text {
    type: "input"
}
```

See [Text][dfn-text].

### `Code`

Inline code.

```idl
interface Code <: Text {
    type: "code"
}
```

See [Text][dfn-text].

### `Paragraph`

A collection of simple text nodes.

```idl
interface Paragraph {
    type: "paragraph"
    children: InlineText[]
}
```

See [InlineText][dfn-inline-text]

### `CodeBlock`

A block of code.

```idl
interface CodeBlock {
    type: "code-block"
    language: "javascript" | "brightscript"
    value: string
}
```

See [Paragraph][dfn-paragraph].

### `Property`

A single prop in the [properties table][dfn-properties].

```idl
interface Property {
    type: "prop"
    prop: Paragraph
    comparator: string
    expectedValue: Paragraph
    actualValue: [Paragraph | CodeBlock]
    status: ["success" | "failure"]
}
``` 

See [Paragraph][dfn-paragraph] and [CodeBlock][dfn-code-block].

### `Properties`

A simple table, where each row represents a property  

```idl
interface Properties {
    type: "props"
    children: Property[]
}
```

See [Property][dfn-property].

### `Condition`

A test line condition. Could be used in assertions or as a part of conditional line.
For example, `element [my element] exists` is a condition and can be used in such lines:

* `Assert element [my element] exists timeout 2s`
* `Press button OK only if element [my element] exists`

Condition consists of title and optional [Properties][dfn-properties] table. Could
also include result status when rendering test line result. 

```idl
interface Condition {
    type: "condition"
    title: Paragraph
    children: [Properties]
    status: ["success" | "fail"]
}
```

See [Paragraph][dfn-paragraph] and [Properties][dfn-properties].

### `TestLine`

Represent a single Suitest test line.

```idl
interface TestLine {
    type: "test-line"
    title: Paragraph
    children: [(Properties | Condition)[]]
    status: ["success" | "fatal" | "fail" | "warning" | "exit" | "excluded"]
}
```

See [Paragraph][dfn-paragraph], [Properties][dfn-properties] and [Condition][dfn-condition]

### `TestLineResult`

Test line result is a wrapper around [TestLine][dfn-test-line], that also includes execution results
for that line.

```idl
interface TestLineResult {
    type: "test-line-result"
    level: "success" | "fatal" | "fail" | "warning" | "exit" | "excluded"
    children: TestLine
    message: [Paragraph],
    screenhost: [string],
}
```

See [Paragraph][dfn-paragraph] and [TestLine][dfn-test-line]

## Enumeration

### `InlineText`

Any inline text node.

```idl
type InlineText = Text | Subject | Input | Code
```

[ext-unist]: https://github.com/syntax-tree/unist
[dfn-text]: #text
[dfn-paragraph]: #paragraph
[dfn-code-block]: #codeblock
[dfn-inline-text]: #inlinetext
[dfn-test-line]: #testline
[dfn-property]: #property
[dfn-properties]: #properties
