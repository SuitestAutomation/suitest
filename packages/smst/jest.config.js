module.exports = {
    "preset": "ts-jest",
    "transform": {
        "^.+\\.(j|t)sx?$": "ts-jest"
    },
    "testMatch": [
        "<rootDir>/lib/__tests__/*.test.ts",
		"<rootDir>/lib/__tests__/*.test.tsx"
    ],
    "collectCoverageFrom": [
        "<rootDir>/lib/**/*.{ts,tsx}",
		"!<rootDir>/lib/__tests__/**/*"
    ],
    "coverageThreshold": {
        "global": {
            "branches": 90,
            "functions": 95,
            "lines": 95,
            "statements": 95
        }
    },
	"moduleFileExtensions": [
		"js", "json", "jsx", "ts", "d.ts", "tsx", "node",
	],
	"globals": {
		"ts-jest": {
			"diagnostics": {
				"warnOnly": true
			},
		},
	},
};
