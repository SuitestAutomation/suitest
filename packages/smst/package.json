{
  "name": "@suitest/smst",
  "version": "3.0.2",
  "description": "UNIST definition and JSX factory for Suitest test lines rendering",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/SuitestAutomation/suitest.git",
    "directory": "packages/smst"
  },
  "kywords": [
    "suitest",
    "unist"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/SuitestAutomation/suitest/issues"
  },
  "homepage": "https://suite.st/",
  "main": "lib/jsxFactory.js",
  "files": [
    "umd/",
    "commonjs/",
    "types/"
  ],
  "scripts": {
    "build": "npm run build:umd && npm run build:commonjs",
    "build:umd": "rimraf umd && rollup -c --environment FORMAT:umd",
    "build:commonjs": "rimraf commonjs && rollup -c --environment FORMAT:commonjs",
    "test": "jest",
    "coverage": "jest --coverage",
    "prepublishOnly": "npm run coverage && npm run build"
  },
  "dependencies": {
    "unist-builder": "^2.0.3"
  },
  "devDependencies": {
    "rollup": "^1.21.4",
    "rimraf": "3.0.0"
  }
}
