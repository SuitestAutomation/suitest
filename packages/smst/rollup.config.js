import typescript from 'rollup-plugin-typescript2';
import { uglify } from "rollup-plugin-uglify";

export default {
	input: './lib/jsxFactory.ts',
	external: [
		'unist-builder',
	],
	plugins: [
		uglify(),
		typescript({tsconfigOverride: {compilerOptions: {module: 'ES2015'}}}),
	],
	output: {
		file: `./${process.env.FORMAT}/jsxFactory.js`,
		name: 'translate',
		format: process.env.FORMAT,
		compact: process.env.FORMAT === 'umd',
		globals: {
			'unist-builder': 'ub',
		},
	},
};
