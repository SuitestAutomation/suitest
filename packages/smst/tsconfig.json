{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "lib",
    "outDir": "commonjs",
    "moduleResolution": "node"
  },
  "include": [
    "lib",
    "types"
  ],
  "exclude": [
    "node_modules",
    "coverage",
    "umd",
    "commonjs",
    "lib/__tests__"
  ]
}
