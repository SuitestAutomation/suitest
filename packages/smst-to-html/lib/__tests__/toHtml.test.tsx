import {jsx} from '@suitest/smst';
import {toHtml, escapeHtml, escapeUrl} from '../toHtml';

describe('AST renderers', () => {
	const defaultOptions = {verbosity: 'normal' as const};
	const plainText = <fragment>TEXT</fragment>;
	const subjectText = <subject>SUBJECT</subject>;
	const inputText = <input>INPUT</input>;
	const codeText = <code>CODE</code>;

	const simpleCodeBlock = <code-block>someJS();</code-block>;
	const longCodeBlock = <code-block>
		someJS(); someJS(); someJS(); someJS();{'\n'}someJS(); someJS(); someJS();{'\n'}someJS(); someJS(); someJS();
	</code-block>;

	const simpleProps = <props>
		<prop
			name={<text>prop name</text>}
			expectedValue={<text>expected value</text>}
			comparator="="
		/>
		<prop
			name={<text>prop name</text>}
			expectedValue={simpleCodeBlock}
			comparator=""
		/>
	</props>;
	const longProps = <props>
		<prop
			name={<text>prop name prop name prop name prop name prop name</text>}
			expectedValue={<text>expected value expected value expected value expected value expected value</text>}
			comparator="~"
			actualValue="expected value expected value expected value expected value expected value"
			status="fail"
		/>
		<prop
			name={<text>JavaScript expression</text>}
			expectedValue={longCodeBlock}
			comparator=""
			status="fail"
		/>
	</props>;

	const simpleCondition = <condition
		title={<fragment>element <subject>My element</subject> exists</fragment>}
	/>;
	const longCondition = <condition
		title={<fragment>element <subject>My element</subject> exists</fragment>}
	>{longProps}</condition>;
	const simpleLine = <test-line
		title={<fragment>Assert element <subject>My element</subject> is visible</fragment>}
	/>;
	const longLine = <test-line title={<fragment>Assert element <subject>My element</subject></fragment>}>
		{simpleCondition}
	</test-line>;

	const failResult = (screenshot?: string): JSX.Element => <test-line-result
		status="fail"
		message={<text>Condition was not met</text>}
		screenshot={screenshot}
		docs="//link/to/docs"
	>
		<test-line
			title={<fragment>Assert element <subject>My element</subject></fragment>}
			status="fail"
		>
			<condition
				title={<fragment>element <subject>My element</subject> exists</fragment>}
				status="fail"
			/>
		</test-line>
	</test-line-result>;
	const warningResult = (screenshot?: string): JSX.Element => <test-line-result
		status="warning"
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="warning"
		/>
	</test-line-result>;
	const exitResult = (screenshot?: string): JSX.Element => <test-line-result
		status="exit"
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="exit"
		/>
	</test-line-result>;
	const excludedResult = (screenshot?: string): JSX.Element => <test-line-result
		status="excluded"
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="excluded"
		/>
	</test-line-result>;
	const fatalResult = (screenshot?: string): JSX.Element => <test-line-result
		status="fatal"
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="fatal"
		/>
	</test-line-result>;
	const abortedResult = (screenshot?: string): JSX.Element => <test-line-result
		status="aborted"
		message={<text>Execution was aborted.</text>}
		screenshot={screenshot}
	>
		<test-line
			title={<fragment>Sleep 10s</fragment>}
			status="aborted"
		/>
	</test-line-result>;

	describe('html renderer', () => {
		it('should handle textual nodes', () => {
			expect(toHtml(plainText, defaultOptions)).toEqual('TEXT');
			expect(toHtml(subjectText, defaultOptions)).toEqual('<span class="suitest-test-line__text--subject">SUBJECT</span>');
			expect(toHtml(inputText, defaultOptions)).toEqual('<span class="suitest-test-line__text--input">INPUT</span>');
			expect(toHtml(codeText, defaultOptions)).toEqual('<code class="suitest-test-line__text--code">CODE</code>');
		});

		it('should handle code blocks', () => {
			expect(toHtml(simpleCodeBlock, defaultOptions)).toMatchSnapshot();
			expect(toHtml(longCodeBlock, defaultOptions)).toMatchSnapshot();
		});

		it('should handle tables', () => {
			expect(toHtml(simpleProps, defaultOptions)).toMatchSnapshot();
			expect(toHtml(longProps, defaultOptions)).toMatchSnapshot();
		});

		it('should render condition', () => {
			expect(toHtml(simpleCondition, defaultOptions)).toMatchSnapshot();
			expect(toHtml(longCondition, defaultOptions)).toMatchSnapshot();
		});

		it('should render test line', () => {
			expect(toHtml(simpleLine, defaultOptions)).toMatchSnapshot();
			expect(toHtml(longLine, defaultOptions)).toMatchSnapshot();
		});

		it('should render alerts', () => {
			expect(toHtml(failResult(), defaultOptions)).toMatchSnapshot();
			expect(toHtml(warningResult(), defaultOptions)).toMatchSnapshot();
			expect(toHtml(exitResult(), defaultOptions)).toMatchSnapshot();
			expect(toHtml(excludedResult(), defaultOptions)).toMatchSnapshot();
			expect(toHtml(fatalResult(), defaultOptions)).toMatchSnapshot();
			expect(toHtml(abortedResult(), defaultOptions)).toMatchSnapshot();
		});

		it('should render results with quiet level', () => {
			expect(toHtml(failResult(), {verbosity: 'quiet'})).toMatchSnapshot();
			expect(toHtml(warningResult(), {verbosity: 'quiet'})).toMatchSnapshot();
			expect(toHtml(exitResult(), {verbosity: 'quiet'})).toMatchSnapshot();
			expect(toHtml(excludedResult(), {verbosity: 'quiet'})).toMatchSnapshot();
			expect(toHtml(fatalResult(), {verbosity: 'quiet'})).toMatchSnapshot();
			expect(toHtml(abortedResult(), { verbosity: 'quiet'})).toMatchSnapshot();
		});

		it('should render test lines with quiet level', () => {
			expect(toHtml(failResult(), {verbosity: 'quiet'})).toMatchSnapshot();
		});

		it('should render test lines with normal level', () => {
			expect(toHtml(failResult(), {verbosity: 'normal'})).toMatchSnapshot();
		});

		it('should render test lines with verbose level', () => {
			expect(toHtml(failResult(), {verbosity: 'verbose'})).toMatchSnapshot();
		});

		it('should render results with screenshots', () => {
			const screenshot = 'path/to/screenshot.png';
			expect(toHtml(failResult(screenshot), defaultOptions)).toMatchSnapshot();
			expect(toHtml(warningResult(screenshot), defaultOptions)).toMatchSnapshot();
			expect(toHtml(exitResult(screenshot), defaultOptions)).toMatchSnapshot();
			expect(toHtml(excludedResult(screenshot), defaultOptions)).toMatchSnapshot();
			expect(toHtml(fatalResult(screenshot), defaultOptions)).toMatchSnapshot();
			expect(toHtml(abortedResult(screenshot), defaultOptions)).toMatchSnapshot();
		});

		it('should render link', () => {
			expect(toHtml(<link href="http://some.url">Some URL</link>, defaultOptions)).toMatchSnapshot();
			expect(toHtml(<link href="http://some.url">http://some.url</link>, defaultOptions)).toMatchSnapshot();
			expect(toHtml(<link href="http://some.url"/>, defaultOptions)).toMatchSnapshot();
		});
	});

	describe('escape html tags', () => {
		const script = '<script>alert("xss")</script>';
		expect(toHtml(<fragment>{script}</fragment>, defaultOptions)).toMatchSnapshot();
		expect(toHtml(<subject>{script}</subject>, defaultOptions)).toMatchSnapshot();
		expect(toHtml(<input>{script}</input>, defaultOptions)).toMatchSnapshot();
		expect(toHtml(<code>{script}</code>, defaultOptions)).toMatchSnapshot();
		expect(toHtml(<code-block>{script}</code-block>, defaultOptions)).toMatchSnapshot();
		expect(toHtml(<text>{script}</text>, defaultOptions)).toMatchSnapshot();
	});

	describe('escape URLs', () => {
		expect(<link href={'?<script>alert("xss")</script>'}>TEST</link>).toMatchSnapshot();
	});

	describe('Translation utils', () => {
		describe('escapeHtml util', () => {
			it('should escape special characters', () => {
				expect(escapeHtml('& < " \'')).toEqual('&amp; &lt; &quot; &#039;');
			});

			it('should replace entities even if they repeat', () => {
				expect(escapeHtml('&&&')).toEqual('&amp;&amp;&amp;');
			});
		});

		describe('escapeUri util', () => {
			it('should escape special characters', () => {
				expect(escapeUrl('& < " \'')).toEqual('&#x0026 &#x003c &#x0022 &#x0027');
			});

			it('should replace entities even if they repeat', () => {
				expect(escapeUrl('&&&')).toEqual('&#x0026&#x0026&#x0026');
			});

			it('should return nothing if URL starts with javascript:', () => {
				expect(escapeUrl('javascript:alert("xss")')).toEqual('');
			});
		});
	});
});
