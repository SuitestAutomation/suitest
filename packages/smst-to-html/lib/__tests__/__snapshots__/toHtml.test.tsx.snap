// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"&lt;script>alert(&quot;xss&quot;)&lt;/script>"`;

exports[` 2`] = `"<span class=\\"suitest-test-line__text--subject\\">&lt;script>alert(&quot;xss&quot;)&lt;/script></span>"`;

exports[` 3`] = `"<span class=\\"suitest-test-line__text--input\\">&lt;script>alert(&quot;xss&quot;)&lt;/script></span>"`;

exports[` 4`] = `"<code class=\\"suitest-test-line__text--code\\">&lt;script>alert(&quot;xss&quot;)&lt;/script></code>"`;

exports[` 5`] = `"<div class=\\"suitest-test-line__code-block\\"><pre><code class=\\"language-javascript\\">&lt;script>alert(&quot;xss&quot;)&lt;/script></code></pre></div>"`;

exports[` 6`] = `"&lt;script>alert(&quot;xss&quot;)&lt;/script>"`;

exports[`AST renderers html renderer should handle code blocks 1`] = `"<div class=\\"suitest-test-line__code-block\\"><pre><code class=\\"language-javascript\\">someJS();</code></pre></div>"`;

exports[`AST renderers html renderer should handle code blocks 2`] = `
"<div class=\\"suitest-test-line__code-block\\"><pre><code class=\\"language-javascript\\">someJS(); someJS(); someJS(); someJS();
someJS(); someJS(); someJS();
someJS(); someJS(); someJS();</code></pre></div>"
`;

exports[`AST renderers html renderer should handle tables 1`] = `"<div class=\\"suitest-test-line__props\\"><table><tr class=\\"suitest-test-line__props__prop--undefined\\"><td>prop name</td><td>=</td><td>expected value</td></tr><tr class=\\"suitest-test-line__props__prop--undefined\\"><td>prop name</td><td></td><td/></tr><tr><td class=\\"suitest-test-line__code-block-prop\\" colspan=\\"3\\"><div class=\\"suitest-test-line__code-block\\"><pre><code class=\\"language-javascript\\">someJS();</code></pre></div></td></tr></table></div>"`;

exports[`AST renderers html renderer should handle tables 2`] = `
"<div class=\\"suitest-test-line__props\\"><table><tr class=\\"suitest-test-line__props__prop--fail\\"><td>prop name prop name prop name prop name prop name</td><td>~</td><td>expected value expected value expected value expected value expected value</td></tr><tr><td/><td>→</td><td>expected value expected value expected value expected value expected value</td></tr><tr class=\\"suitest-test-line__props__prop--fail\\"><td>JavaScript expression</td><td></td><td/></tr><tr><td class=\\"suitest-test-line__code-block-prop\\" colspan=\\"3\\"><div class=\\"suitest-test-line__code-block\\"><pre><code class=\\"language-javascript\\">someJS(); someJS(); someJS(); someJS();
someJS(); someJS(); someJS();
someJS(); someJS(); someJS();</code></pre></div></td></tr></table></div>"
`;

exports[`AST renderers html renderer should render alerts 1`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--fail\\"><div class=\\"suitest-test-line suitest-test-line--fail\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div><div class=\\"suitest-test-line__result__message\\">Condition was not met</div></div>"`;

exports[`AST renderers html renderer should render alerts 2`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--warning\\"><div class=\\"suitest-test-line suitest-test-line--warning\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div></div>"`;

exports[`AST renderers html renderer should render alerts 3`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--exit\\"><div class=\\"suitest-test-line suitest-test-line--exit\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div></div>"`;

exports[`AST renderers html renderer should render alerts 4`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--excluded\\"><div class=\\"suitest-test-line suitest-test-line--excluded\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div></div>"`;

exports[`AST renderers html renderer should render alerts 5`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--fatal\\"><div class=\\"suitest-test-line suitest-test-line--fatal\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div></div>"`;

exports[`AST renderers html renderer should render condition 1`] = `"<div class=\\"suitest-test-line__condition suitest-test-line__condition--undefined\\"><div class=\\"suitest-test-line__condition suitest-test-line__condition--undefined__header\\">condition: condition: element <span class=\\"suitest-test-line__text--subject\\">My element</span> exists</div></div>"`;

exports[`AST renderers html renderer should render condition 2`] = `
"<div class=\\"suitest-test-line__condition suitest-test-line__condition--undefined\\"><div class=\\"suitest-test-line__condition suitest-test-line__condition--undefined__header\\">condition: condition: element <span class=\\"suitest-test-line__text--subject\\">My element</span> exists</div><div class=\\"suitest-test-line__props\\"><table><tr class=\\"suitest-test-line__props__prop--fail\\"><td>prop name prop name prop name prop name prop name</td><td>~</td><td>expected value expected value expected value expected value expected value</td></tr><tr><td/><td>→</td><td>expected value expected value expected value expected value expected value</td></tr><tr class=\\"suitest-test-line__props__prop--fail\\"><td>JavaScript expression</td><td></td><td/></tr><tr><td class=\\"suitest-test-line__code-block-prop\\" colspan=\\"3\\"><div class=\\"suitest-test-line__code-block\\"><pre><code class=\\"language-javascript\\">someJS(); someJS(); someJS(); someJS();
someJS(); someJS(); someJS();
someJS(); someJS(); someJS();</code></pre></div></td></tr></table></div></div>"
`;

exports[`AST renderers html renderer should render results with screenshots 1`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--fail\\"><div class=\\"suitest-test-line suitest-test-line--fail\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div><div class=\\"suitest-test-line__result__message\\">Condition was not met</div><div class=\\"suitest-test-line__result__screenshot\\">screenshot: <a href=\\"path/to/screenshot.png\\" target=\\"_blank\\">path/to/screenshot.png</a></div></div>"`;

exports[`AST renderers html renderer should render results with screenshots 2`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--warning\\"><div class=\\"suitest-test-line suitest-test-line--warning\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div><div class=\\"suitest-test-line__result__screenshot\\">screenshot: <a href=\\"path/to/screenshot.png\\" target=\\"_blank\\">path/to/screenshot.png</a></div></div>"`;

exports[`AST renderers html renderer should render results with screenshots 3`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--exit\\"><div class=\\"suitest-test-line suitest-test-line--exit\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div><div class=\\"suitest-test-line__result__screenshot\\">screenshot: <a href=\\"path/to/screenshot.png\\" target=\\"_blank\\">path/to/screenshot.png</a></div></div>"`;

exports[`AST renderers html renderer should render results with screenshots 4`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--excluded\\"><div class=\\"suitest-test-line suitest-test-line--excluded\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div><div class=\\"suitest-test-line__result__screenshot\\">screenshot: <a href=\\"path/to/screenshot.png\\" target=\\"_blank\\">path/to/screenshot.png</a></div></div>"`;

exports[`AST renderers html renderer should render results with screenshots 5`] = `"<div class=\\"suitest-test-line__result suitest-test-line__result--fatal\\"><div class=\\"suitest-test-line suitest-test-line--fatal\\"><div class=\\"suitest-test-line__title\\">Assert application has exited</div></div><div class=\\"suitest-test-line__result__screenshot\\">screenshot: <a href=\\"path/to/screenshot.png\\" target=\\"_blank\\">path/to/screenshot.png</a></div></div>"`;

exports[`AST renderers html renderer should render test line 1`] = `"<div class=\\"suitest-test-line suitest-test-line--undefined\\"><div class=\\"suitest-test-line__title\\">Assert element <span class=\\"suitest-test-line__text--subject\\">My element</span> is visible</div></div>"`;

exports[`AST renderers html renderer should render test line 2`] = `"<div class=\\"suitest-test-line suitest-test-line--undefined\\"><div class=\\"suitest-test-line__title\\">Assert element <span class=\\"suitest-test-line__text--subject\\">My element</span></div><div class=\\"suitest-test-line__condition suitest-test-line__condition--undefined\\"><div class=\\"suitest-test-line__condition suitest-test-line__condition--undefined__header\\">condition: condition: element <span class=\\"suitest-test-line__text--subject\\">My element</span> exists</div></div></div>"`;
